### DISCUSSION TITLE
Explanation of code design/logic

### DISCUSSION THREAD

**Discussion by manickavela29** [k2-fsa/sherpa-onnx#980] [Q&A] [answered]
Hi @csukuangfj , @pkufool 

It would be great if someone can explain the idea behind unlocking the mutex before DecodeStreams and locking it afterwards?

https://github.com/k2-fsa/sherpa-onnx/blob/1a43d1e37f2a65a7326e75be4607b4996f9737a8/sherpa-onnx/csrc/online-websocket-server-impl.cc#L201

as per my understanding,
 several instances of DecodeStreams can run concurrently with different sets of s_vec, is it right?

**Comment by csukuangfj** [comment]
~~The lock is to protect `s_vec`~~

---

> several instances of DecodeStreams can run concurrently with different sets of s_vec, is it right?

Yes, you are right.

Recognizer is thread-safe while OnlineStream is not thread safe.

Several threads can use the same recognizer to decode different streams.

The lock protects data members such as `ready_connections_`.

**Reply by manickavela29** [reply]
Thanks for the explanation
How does onnxruntime inference behave with this,

does this mean several batch inference can happen at same time on onnxruntime session ?
or is it kind of queued for onnxruntime to infer?

since DecodeStreams is unlocked, several thread(with unique set of streams under active list) might enter Decodestreams together,  how is the onnxrt session resources shared between these, or do they happen sequentially

**Comment by manickavela29** [comment]
Hi @csukuangfj ,

> Thanks for the explanation How does onnxruntime inference behave with this,
> 
> does this mean several batch inference can happen at same time on onnxruntime session ? or is it kind of queued for onnxruntime to infer?
> 
> since DecodeStreams is unlocked, several thread(with unique set of streams under active list) might enter Decodestreams together, how is the onnxrt session resources shared between these, or do they happen sequentially

**Reply by csukuangfj** [reply]
Please see https://github.com/microsoft/onnxruntime/issues/114

onnxruntime session is thread-safe.

----

> , or do they happen sequentially

No, they run concurrently and independently between threads.

**Reply by csukuangfj** [reply, marked as answer]
By the way, from my experience, onnxruntime session is thread-safe when you use CPU.
It is not the case when you use CUDA.

**Reply by manickavela29** [reply]
Thanks,  I had the very same question with CUDA and was wondering if it was handled.
but it seems to be working fine 😄, so all good then.

so this concurrency strategy is specific to CPU which is just extended to CUDA(TRT too) right?

`recognizer_->DecodeStreams(s_vec.data(), s_vec.size());
`

**Reply by csukuangfj** [reply]
> so this concurrency strategy is specific to CPU which is just extended to CUDA(TRT too) right?



Everything should be fine if it does not throw errors when you switch to CUDA/TensorRT.
Otherwise, you have to hold the lock.
